File: apps/account/src/main/kotlin/io/github/songminkyu/account/logging/restclient/LoggingInterceptor.java
package io.github.songminkyu.account.logging.restclient;

import io.github.songminkyu.account.logging.core.HttpResponse;
import io.github.songminkyu.account.logging.core.Sink;
import java.io.IOException;
import java.io.UncheckedIOException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpRequestExecution;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.ClientHttpResponse;
import org.jspecify.annotations.NullMarked;

@Slf4j
@RequiredArgsConstructor
@NullMarked
public class LoggingInterceptor implements ClientHttpRequestInterceptor {

    private final Sink sink;
    
    @Override
    public ClientHttpResponse intercept(
        HttpRequest clientRequest,
        byte[] body,
        ClientHttpRequestExecution execution) {
        try {
            logRequest(clientRequest, body);
            ClientHttpResponse response = execution.execute(clientRequest, body);
            logResponse(response);
            return response;
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private void logRequest(HttpRequest request, byte[] body) throws IOException {
        final io.github.songminkyu.account.logging.core.HttpRequest httpRequest =
            new RestClientRequest(request, body);
        sink.write(httpRequest);
    }

    private void logResponse(ClientHttpResponse response) throws IOException {
        HttpResponse clientResponse = new RestClientResponse(response);
        sink.write(clientResponse);
    }
}

File: backup/account-kotlin/src/main/kotlin/io/github/songminkyu/account/logging/restclient/LoggingInterceptor.kt
package io.github.songminkyu.account.logging.restclient

import io.github.oshai.kotlinlogging.KotlinLogging
import io.github.songminkyu.account.logging.core.HttpResponse
import io.github.songminkyu.account.logging.core.Sink
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse
import java.io.IOException
import java.io.UncheckedIOException

private val logger = KotlinLogging.logger {}

class LoggingInterceptor(
    private val sink: Sink
) : ClientHttpRequestInterceptor {

    override fun intercept(
        clientRequest: HttpRequest,
        body: ByteArray,
        execution: ClientHttpRequestExecution
    ): ClientHttpResponse {
        return try {
            logRequest(clientRequest, body)
            val response = execution.execute(clientRequest, body)
            logResponse(response)
            response
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }
    }

    @Throws(IOException::class)
    private fun logRequest(request: HttpRequest, body: ByteArray) {
        val httpRequest = RestClientRequest(request, body)
        sink.write(httpRequest)
    }

    @Throws(IOException::class)
    private fun logResponse(response: ClientHttpResponse) {
        val clientResponse: HttpResponse = RestClientResponse(response)
        sink.write(clientResponse)
    }
}